src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/NumericRangeRewriter.cs

```diff
@@ -0,0 +1,73 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// Rewrites expressions over Quantity and Number values to take ranges into account.
+    /// They will combine the results of the original expression with the results of the
+    /// expressions over the low and high fields where the entry is not a single value.
+    /// </summary>
+    internal class NumericRangeRewriter : ConcatenationRewriter
+    {
+        internal static readonly NumericRangeRewriter Instance = new NumericRangeRewriter();
+
+        private NumericRangeRewriter()
+            : base(new Scout())
+        {
+        }
+
+        public override Expression VisitBinary(BinaryExpression expression, object context)
+        {
+            FieldName highField;
+            FieldName lowField;
+            switch (expression.FieldName)
+            {
+                case FieldName.Quantity:
+                    highField = SqlFieldName.QuantityHigh;
+                    lowField = SqlFieldName.QuantityLow;
+                    break;
+                case FieldName.Number:
+                    highField = SqlFieldName.NumberHigh;
+                    lowField = SqlFieldName.NumberLow;
+                    break;
+                default:
+                    return expression;
+            }
+
+            switch (expression.BinaryOperator)
+            {
+                case BinaryOperator.GreaterThan:
+                    return Expression.GreaterThan(highField, expression.ComponentIndex, expression.Value);
+                case BinaryOperator.GreaterThanOrEqual:
+                    return Expression.GreaterThanOrEqual(highField, expression.ComponentIndex, expression.Value);
+                case BinaryOperator.LessThan:
+                    return Expression.LessThan(lowField, expression.ComponentIndex, expression.Value);
+                case BinaryOperator.LessThanOrEqual:
+                    return Expression.LessThanOrEqual(lowField, expression.ComponentIndex, expression.Value);
+                case BinaryOperator.Equal:
+                case BinaryOperator.NotEqual:
+                default:
+                    throw new ArgumentOutOfRangeException(expression.BinaryOperator.ToString());
```

Comment:
Wouldn't these fall into the following parts of the spec (https://www.hl7.org/fhir/search.html#prefix): 

|  |  |  |
| ---  | ---   | --- |
| eq | the value for the parameter in the resource is equal to the provided value | the range of the search value fully contains the range of the target value |
| ne | the value for the parameter in the resource is not equal to the provided value | the range of the search value does not fully contain the range of the target value |

Reply:
The Core layer, when creating these expressions, turns all equality expressions into range expressions, so we do not need to handle equality here.